### PR TITLE
[dtensor] skip pytree when not necessary

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110132

pytree is a great tool, but it sometimes considers to be evil for
tensor subclasses, it's useful to implement subclass quickly, but it:
* exposes non-trival CPU overhead
* many ops don't need pytree, only the one with list/dict ops needs
* blindly use pytree to re-wrap have semantic issues for inplace/out
ops

This PR avoid using pytree for most ops during torch_dispatch and only
enable it for certain ops

### CODE REVIEW

File: torch/distributed/_tensor/dispatch.py [hunk 1]
```diff
@@ -82,24 +82,29 @@ def redistribute_local_args(
...
-                new_flat_local_args.append(resharded_local_tensor)
+                new_local_args.append(resharded_local_tensor)
             else:
-                new_flat_local_args.append(local_tensor)
+                new_local_args.append(local_tensor)
         else:
-            new_flat_local_args.append(reshard_arg_spec)
+            new_local_args.append(reshard_arg_spec)
 
-    op_info.flat_local_args = new_flat_local_args
```
Comment (fduwjj) [inline torch/distributed/_tensor/dispatch.py:102]: it's interesting it worked without `tuple()` previously..

**Review by fduwjj — APPROVED** [review]
LGTM, it's actually surprising to me that only two ops need treemap here..

File: torch/distributed/_tensor/dispatch.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:102]: Yeah it's because we always to unflatten before, which recovers the list to tuple

**Comment by Chillee** [comment]
Have you tried using optree?

**Comment by wanchaol** [comment]
> Have you tried using optree?

Oh just noticed that got landed!! I can try turn that on, is there a flag to turn it on? 

My feeling is that even if that landed it would still have some overhead exposed (especially optree perf is most amazing on dicts, while operator arguments does not have too many dicts, maybe kwargs is a good try). This PR try to turn pytree/optree off completely for most ops which I thought it would be faster for either case?

**Comment by Chillee** [comment]
@wanchaol might be a little bit annoying to actually land but the idea was that we would incrementally port usages over incrementally.

**Comment by wanchaol** [comment]
> @wanchaol might be a little bit annoying to actually land but the idea was that we would incrementally port usages over incrementally.

@Chillee sounds good, I'll give it a try in a follow up PR? the annoying part is that we need to guard the imports mostly so that all CI passes?

**Comment by Chillee** [comment]
Yeah, see https://github.com/pytorch/pytorch/pull/109684

**Pull request closed without merging**